conda/core/index.py

```diff
@@ -328,7 +332,65 @@ def maybe_decompress(filename, resp_content):
                              getattr(e.response, 'elapsed', None))
 
 
-def read_local_repodata(cache_path):
+def write_pickled_repodata(cache_path, repodata):
+    # Don't bother to pickle empty channels
+    if not repodata.get('packages'):
+        return
+    try:
+        with open(cache_path + '.q', 'wb') as f:
+            pickle.dump(repodata, f)
+    except Exception as e:
+        import traceback
+        log.debug("Failed to dump pickled repodata.\n%s", traceback.format_exc())
+
+
+def read_pickled_repodata(cache_path, channel_url, schannel, priority):
+    pickle_path = cache_path + '.q'
+    # Don't trust pickled data if there is no accompanying json data
+    if not isfile(pickle_path) or not isfile(cache_path):
+        return None
+    try:
+        with open(cache_path + '.q', 'rb') as f:
+            repodata = pickle.load(f)
+    except Exception as e:
+        import traceback
+        log.debug("Failed to load pickled repodata.\n%s", traceback.format_exc())
+        return None
+    if repodata.get('_url') != channel_url:
+        return None
+    add_pip = context.add_pip_as_python_dependency
+    changed = False
+    # Safe changes to IndexRecord
+    if ((repodata['_add_pip'] != add_pip or
+         repodata['_priority'] != priority or
+         repodata['_schannel'] != schannel)):
+        repodata['_priority'] = priority
+        repodata['_add_pip'] = add_pip
+        changed = True
+        for rec in itervalues(repodata.get('packages', {})):
+            rec.priority = priority
+            rec.schannel = schannel
+            if rec.name == 'python' and rec.version.startswith(('2.', '3.')):
+                if add_pip:
+                    rec.depends = rec.depends + ('pip',)
+                elif rec.depends[-1] == 'pip':
+                    rec.depends = rec.depends[:-1]
+    # Safe changes to Dist
+    if repodata['_schannel'] != schannel:
+        repodata['_schannel'] = schannel
+        repodata['packages'] = {
+            Dist.from_string(dist.fn, channel_override=schannel): rec
+            for dist, rec in itervalues(repodata.get('packages', {}))}
+        changed = True
+    if changed:
+        write_pickled_repodata(cache_path, repodata)
+    return repodata
+
+
+def read_local_repodata(cache_path, channel_url, schannel, priority):
+    local_repodata = read_pickled_repodata(cache_path, channel_url, schannel, priority)
```

Comment:
This `local_repodata` is already going to have `_mod` and `_etag` keys.  As an extra level of caution, I'd propose passing those values through as arguments to this function, and then to `read_pickled_repodata`, so on line 359

    if repodata.get('_url') != channel_url:

we can be extra cautious that our pickled data logic is in-sync with everything else.